File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/shared/MagicPropertyBuilder.java
package com.lasthopesoftware.bluewater.shared;

/**
 * Created by david on 7/24/15.
 */
public class MagicPropertyBuilder {
	public static String buildMagicPropertyName(Class c, String propertyName) {
		return buildMagicPropertyName(c.getCanonicalName(), propertyName);
	}

	private final String canonicalName;

	public MagicPropertyBuilder(Class c) {
		this.canonicalName = c.getCanonicalName();
	}

	public String buildProperty(String propertyName) {
		return buildMagicPropertyName(canonicalName, propertyName);
	}

	private static String buildMagicPropertyName(String prefix, String propertyName) {
		return prefix + "." + propertyName;
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/shared/MagicPropertyBuilder.kt
package com.lasthopesoftware.bluewater.shared

class MagicPropertyBuilder(c: Class<*>) {
	private val canonicalName = c.canonicalName

	fun buildProperty(propertyName: String): String = buildMagicPropertyName(canonicalName, propertyName)

	companion object {
		inline fun <reified T> buildMagicPropertyName(propertyName: String) =
			buildMagicPropertyName(T::class.java, propertyName)

		@JvmStatic
		fun buildMagicPropertyName(c: Class<*>, propertyName: String): String =
			buildMagicPropertyName(c.canonicalName, propertyName)

		private fun buildMagicPropertyName(prefix: String?, propertyName: String): String =
			"$prefix.$propertyName"
	}
}
